fv3core/utils/grid.py

```diff
@@ -366,14 +367,202 @@ def horizontal_starts_from_shape(self, shape):
+    ):
+        """
+        Initialize a grid indexing object.
+
+        Args:
+            origin: index of the compute domain origin for cell-centered variables
+            domain: size of the compute domain for cell-centered variables
+            n_halo: number of halo points
+            south_edge: whether the current rank is on the south edge of a tile
+            north_edge: whether the current rank is on the north edge of a tile
+            west_edge: whether the current rank is on the west edge of a tile
+            east_edge: whether the current rank is on the east edge of a tile
+        """
+        self.origin = origin
+        self.domain = domain
+        self.n_halo = n_halo
+        self.south_edge = south_edge
+        self.north_edge = north_edge
+        self.west_edge = west_edge
+        self.east_edge = east_edge
+
+    @classmethod
+    def from_sizer_and_communicator(
+        cls, sizer: fv3util.GridSizer, cube: fv3util.CubedSphereCommunicator
+    ) -> "GridIndexing":
+        # TODO: if this class is refactored to split off the *_edge booleans,
+        # this init routine can be refactored to require only a GridSizer
+        origin = sizer.get_origin([fv3util.X_DIM, fv3util.Y_DIM, fv3util.Z_DIM])
+        domain = sizer.get_extent([fv3util.X_DIM, fv3util.Y_DIM, fv3util.Z_DIM])
+        south_edge = cube.tile.on_tile_bottom(cube.rank)
+        north_edge = cube.tile.on_tile_top(cube.rank)
+        west_edge = cube.tile.on_tile_left(cube.rank)
+        east_edge = cube.tile.on_tile_right(cube.rank)
+        return cls(
+            origin=origin,
+            domain=domain,
+            n_halo=sizer.n_halo,
+            south_edge=south_edge,
+            north_edge=north_edge,
+            west_edge=west_edge,
+            east_edge=east_edge,
+        )
+
+    @classmethod
+    def from_legacy_grid(cls, grid: Grid) -> "GridIndexing":
+        return cls(
+            origin=grid.compute_origin(),
+            domain=grid.domain_shape_compute(),
+            n_halo=grid.halo,
+            south_edge=grid.south_edge,
+            north_edge=grid.north_edge,
+            west_edge=grid.west_edge,
+            east_edge=grid.east_edge,
+        )
+
+    @property
+    def isc(self):
+        """start of the compute domain along the x-axis"""
+        return self.origin[0]
+
+    @property
+    def iec(self):
+        """end of the compute domain along the x-axis"""
+        return self.origin[0] + self.domain[0]
+
+    @property
+    def jsc(self):
+        """start of the compute domain along the y-axis"""
+        return self.origin[1]
+
+    @property
+    def jec(self):
+        """end of the compute domain along the y-axis"""
+        return self.origin[1] + self.domain[1]
+
+    @property
+    def ks(self):
+        """start of the data domain along the z-axis"""
+        return self.origin[2]
+
+    @property
+    def ke(self):
+        """end of the data domain along the z-axis"""
```

Comment:
It may be OK to "duplicate" ks/ke in that way, in the same way that there were modules that took in "istart" or "iend". In both cases they have the same meaning (the start and end indices in the vertical), just applied to the appropriate context for the object (GridIndexing is about the full grid, RayleighFriction is about where you're applying Rayleigh friction).

> This may result in someone initializing a new grid indexer when they are using different values there, which we probably don't want

Yes, definitely. We talked about ways to handle vertical indexing separately and use GridIndexing for just the horizontal. Hopefully this doesn't come up often. If we find there are a lot of cases that need that then we can think about how to integrate those special cases into the GridIndexing object.